massa-execution/src/types.rs

```diff
@@ -24,6 +29,59 @@ pub(crate) struct StepHistoryItem {
 
     // list of SCE ledger changes caused by this execution step
     pub ledger_changes: SCELedgerChanges,
+
+    pub events: EventStore,
+}
+
+#[derive(Default, Debug, Clone)]
+pub(crate) struct EventStore {
+    id_to_event: Map<SCOutputEventId, SCOutputEvent>,
+    slot_to_id: HashMap<Slot, Set<SCOutputEventId>>,
+    caller_to_id: Map<Address, Set<SCOutputEventId>>,
+    smart_contract_to_id: Map<Address, Set<SCOutputEventId>>,
+}
+
+impl EventStore {
+    // add event to the store and all its indexes
+    pub fn insert(&mut self, id: SCOutputEventId, event: SCOutputEvent) {
+        if let Entry::Vacant(entry) = self.id_to_event.entry(id) {
+            match self.slot_to_id.entry(event.context.slot) {
+                Entry::Occupied(mut e) => {
+                    e.get_mut().insert(id);
+                }
+                Entry::Vacant(e) => {
+                    e.insert(vec![id].into_iter().collect());
+                }
+            }
+            if let Some(caller) = event.context.call_stack.front() {
+                self.caller_to_id
+                    .get_mut(caller)
+                    .unwrap_or(&mut vec![].into_iter().collect()) // TODO maybe implement default for Set
+                    .insert(id);
+            }
```

Comment:
In a nutshell here I do the same thing twice (but for a different map) and I'm not sure which way is the better any opinions @yvan-sraka @adrien-zinger

Reply:
Thanks for your help guys ! :)